---
title: "Alignment using Bowtie2"
author: Tamas Schauer
date: '`r format(Sys.time(), "%d %B, %Y")`'
output: 
    html_document:
        toc: true
        toc_depth: 3
---
<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Quick Intro to Command Line

* Open the Terminal app from /Applications/Utilities on your Mac
* Check the following commands

```{sh, comment=NA, message=FALSE, warning=FALSE, results='hide', eval=FALSE}
# change to home directory
cd ~

# print working directory
pwd

# make new directory
mkdir Software

# list files
ls -l

# manual page of cp command
man cp
```


## Install Bowtie2

* Bowtie2 executables are available on sourceforge
* download and extract the files

```{bash, comment=NA, message=FALSE, warning=FALSE, results='hide', eval=FALSE}
# change to Software directory
cd ~/Software

# download from sourceforge
curl -LO https://sourceforge.net/projects/bowtie-bio/files/bowtie2/2.4.1/bowtie2-2.4.1-macos-x86_64.zip

# extract zip file
unzip bowtie2-2.4.1-macos-x86_64.zip
```

### Make Bowtie2 available for Terminal


### Add to bash_profile

```{bash, eval=FALSE}
# start from the home directory
cd ~   

# create .bash_profile (if you donÂ´t have it)
touch .bash_profile

# open .bash_profile in Atom (or which ever text editor)
open -a Atom .bash_profile

# add the following line with your path of Bowtie2 to the .bash_profile file
PATH=/Users/tschauer/Software/bowtie2-2.4.1-macos-x86_64:$PATH
# save and close window

# source .bash_profile
source .bash_profile

# print .bash_profile
cat .bash_profile
```


## Install Samtools

```{bash, comment=NA, message=FALSE, warning=FALSE, results='hide', eval=FALSE}
# change to Software directory
cd ~/Software

# download from sourceforge
curl -LO https://github.com/samtools/samtools/releases/download/1.10/samtools-1.10.tar.bz2

# extract tar.bz2 file
tar -xjf samtools-1.10.tar.bz2 

# build from source
cd samtools-1.10
./configure --prefix=/Users/tschauer/Software/samtools-1.10
make
make install


# add the following line with your path of Bowtie2 to the .bash_profile file
PATH=/Users/tschauer/Software/samtools-1.10:$PATH
# save and close window

# source .bash_profile
source .bash_profile


```

```{bash, comment=NA, eval=FALSE}
# start from the home directory
cd ~   

# get first few lines of help page
bowtie2 --help  | head
samtools --help | head
```

## Create Bowtie2 Index 

```{bash, eval=FALSE}
# start from the home directory
cd ~   

# make new directory for genomes
mkdir Genomes

# change directory to genomes
cd Genomes/

# make new directory for Drosophila
mkdir Dmelanogaster

# change directory to genomes
cd Dmelanogaster/

# Download genome fasta file from database
curl -LO ftp://ftp.flybase.net/genomes/Drosophila_melanogaster/dmel_r6.17_FB2017_04/fasta/dmel-all-chromosome-r6.17.fasta.gz

# Extract fasta file
gunzip dmel-all-chromosome-r6.17.fasta.gz

# rename chromosome names with chr prefix using sed command (optional)
samtools faidx dmel-all-chromosome-r6.17.fasta 2L 2R 3L 3R 4 X Y | sed 's/^>/>chr/' > dmel-all-chromosome-r6.17.fa

# create bowtie2 index
bowtie2-build dmel-all-chromosome-r6.17.fa dmel_genome
```

## Run Bowtie2

```{bash, eval=FALSE}
# start from the home directory
cd ~   

# make new directory structure for your project and subdirectories
mkdir -p Projects/Project02/FastQ

cd Projects/Project02/FastQ

# move files from Downloads folder to your Project folder
mv /Users/tschauer/Downloads/H2AvChIP_*.gz .

# run bowtie2 (See how to automatize this in an other tutorial)
bowtie2 -p 2 -x  ~/Genomes/Dmelanogaster/dmel_genome -U H2AvChIP_GFP_INP_1.txt.gz  > H2AvChIP_GFP_INP.sam 2> H2AvChIP_GFP_INP.txt
bowtie2 -p 2 -x  ~/Genomes/Dmelanogaster/dmel_genome -U H2AvChIP_GFP_IP_1.txt.gz   > H2AvChIP_GFP_IP.sam  2> H2AvChIP_GFP_IP.txt

```


## Check Input and Output files

```{bash}
# path is from home directory (~)
head ~/Genomes/Dmelanogaster/dmel-all-chromosome-r6.17.fa
```

```{bash}
# change to project directory
cd ~/Projects/Project02/FastQ/

zcat < H2AvChIP_GFP_INP_1.txt.gz | head -n 20
```

```{bash, eval=FALSE}
head -n 20 H2AvChIP_GFP_INP.sam
```

```{bash, echo=FALSE}
cd ~/Projects/Project02/FastQ/
head -n 20 H2AvChIP_GFP_INP.sam
```


```{bash, eval=FALSE}
cat H2AvChIP_GFP_INP.txt
```

```{bash, echo=FALSE}
cd ~/Projects/Project02/FastQ/
cat H2AvChIP_GFP_INP.txt
```






